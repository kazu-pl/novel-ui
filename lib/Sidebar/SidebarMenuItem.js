import { _ as __makeTemplateObject, a as __assign } from '../tslib.es6-905b4068.js';
import { jsxs, Fragment, jsx } from 'react/jsx-runtime';
import Collapse from '@mui/material/Collapse';
import List from '@mui/material/List';
import ExpandLess from '@mui/icons-material/ExpandLess';
import ExpandMore from '@mui/icons-material/ExpandMore';
import { useState } from 'react';
import ColoredIconWrapper from '../ColoredIconWrapper/ColoredIconWrapper.js';
import styled, { css } from 'styled-components';
import ListItemText from '@mui/material/ListItemText';
import ListItemIcon from '@mui/material/ListItemIcon';
import { NavLink, useLocation } from 'react-router-dom';
import Divider from '@mui/material/Divider';
import { alpha } from '@mui/material/styles';

var commonLinkStyles = css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  width: 100%;\n  display: flex;\n  justify-content: flex-start;\n  align-items: center;\n  text-align: left;\n  box-sizing: border-box;\n  text-decoration: none;\n  color: ", ";\n"], ["\n  width: 100%;\n  display: flex;\n  justify-content: flex-start;\n  align-items: center;\n  text-align: left;\n  box-sizing: border-box;\n  text-decoration: none;\n  color: ", ";\n"])), function (_a) {
    var theme = _a.theme;
    return theme.palette.text.primary;
});
var hoverLinkStyles = css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  background-color: ", ";\n  border-radius: 8px;\n"], ["\n  background-color: ", ";\n  border-radius: 8px;\n"])), function (_a) {
    var theme = _a.theme;
    return alpha(theme.palette.primary.light, 0.1);
});
var activeLinkStyles = css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  background-color: ", ";\n  border-radius: 8px;\n"], ["\n  background-color: ", ";\n  border-radius: 8px;\n"])), function (_a) {
    var theme = _a.theme;
    return alpha(theme.palette.primary.light, 0.2);
});
var StyledListItemText = styled(ListItemText)(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n  color: ", ";\n"], ["\n  color: ", ";\n"])), function (_a) {
    var theme = _a.theme;
    return theme.palette.text.primary;
});
var StyledListItemIcon = styled(ListItemIcon)(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n  min-width: 40px;\n"], ["\n  min-width: 40px;\n"])));
var StyledLink = styled(NavLink)(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  ", "\n  padding: 8px 16px 8px ", "px;\n\n  &:hover {\n    ", "\n  }\n\n  &.active {\n    ", "\n  }\n"], ["\n  ", "\n  padding: 8px 16px 8px ", "px;\n\n  &:hover {\n    ", "\n  }\n\n  &.active {\n    ", "\n  }\n"])), commonLinkStyles, function (_a) {
    var $asSubmenuItem = _a.$asSubmenuItem;
    return ($asSubmenuItem ? 48 : 8);
}, hoverLinkStyles, activeLinkStyles);
var StyledListItemButton = styled.button(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  ", "\n  padding: 8px;\n  cursor: pointer;\n  background: none;\n  border: none;\n\n  &:hover {\n    ", "\n  }\n"], ["\n  ", "\n  padding: 8px;\n  cursor: pointer;\n  background: none;\n  border: none;\n\n  &:hover {\n    ", "\n  }\n"])), commonLinkStyles, hoverLinkStyles);
var StyledDividerWrapper = styled.div(templateObject_8 || (templateObject_8 = __makeTemplateObject(["\n  padding: 8px 0;\n"], ["\n  padding: 8px 0;\n"])));
var StyledDivider = styled(Divider)(templateObject_9 || (templateObject_9 = __makeTemplateObject(["\n  border-color: ", ";\n"], ["\n  border-color: ", ";\n"])), function (_a) {
    var theme = _a.theme;
    return theme.palette.grey[300];
});
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9;

var SidebarMenuItem = function (props) {
    var location = useLocation();
    var isSubmenuExpandedByDefault = props.variant === "with-dropdown"
        ? props.dropdownItems.some(function (item) { return location.pathname.includes(item.to); })
        : false;
    var _a = useState(isSubmenuExpandedByDefault), isSubmenuOpen = _a[0], setIsSubmenuOpen = _a[1];
    var handleClick = function () { return setIsSubmenuOpen(function (prev) { return !prev; }); };
    return props.variant === "with-dropdown" ? (jsxs(Fragment, { children: [jsxs(StyledListItemButton, __assign({ onClick: handleClick }, { children: [jsx(StyledListItemIcon, { children: props.icon }, void 0), jsx(StyledListItemText, { primary: props.label, color: "#ffffff" }, void 0), isSubmenuOpen ? (jsx(ColoredIconWrapper, __assign({ color: "grey" }, { children: jsx(ExpandLess, {}, void 0) }), void 0)) : (jsx(ColoredIconWrapper, __assign({ color: "grey" }, { children: jsx(ExpandMore, {}, void 0) }), void 0))] }), void 0), jsx(Collapse, __assign({ in: isSubmenuOpen, timeout: "auto", unmountOnExit: true }, { children: jsx(List, __assign({ component: "div", disablePadding: true }, { children: props.dropdownItems.map(function (item, index) { return (jsx(StyledLink, __assign({ to: item.to, "$asSubmenuItem": true }, { children: item.label }), index)); }) }), void 0) }), void 0), props.renderBottomLine && (jsx(StyledDividerWrapper, { children: jsx(StyledDivider, {}, void 0) }, void 0))] }, void 0)) : (jsxs(Fragment, { children: [jsxs(StyledLink, __assign({ to: props.to }, { children: [jsx(StyledListItemIcon, { children: props.icon }, void 0), jsx(StyledListItemText, { primary: props.label }, void 0)] }), void 0), props.renderBottomLine && (jsx(StyledDividerWrapper, { children: jsx(StyledDivider, {}, void 0) }, void 0))] }, void 0));
};

export { SidebarMenuItem as default };
