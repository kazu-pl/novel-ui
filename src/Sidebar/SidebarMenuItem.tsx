import Collapse from "@mui/material/Collapse";
import List from "@mui/material/List";
import ExpandLess from "@mui/icons-material/ExpandLess";
import ExpandMore from "@mui/icons-material/ExpandMore";
import { useState } from "react";
import ColoredIconWrapper from "../../src/ColoredIconWrapper";

import {
  StyledListItemText,
  StyledListItemIcon,
  StyledLink,
  StyledListItemButton,
  StyledDividerWrapper,
  StyledDivider,
} from "./SidebarMenuItem.styled";
import { useLocation } from "react-router-dom";

export const bottomLineWrapperTestId = "bottomLineWrapperTestId";
/**
 * It's used for icon when dropdown is expanded so this icon indicates that when clicked the dropdown would collapse
 */
export const expandLessIconTestId = "expandLessIconTestId";
/**
 * It's used for icon when dropdown is collapsed so this icon indicates that when clicked the dropdown would expand
 */
export const expandMoreIconTestId = "expandMoreIconTestId";
export const dropdownCollapsableWrapper = "dropdownCollapsableWrapper";

export interface SidebarMenuItemWithoutDropdown {
  variant: "no-dropdown";
  icon: React.ReactNode;
  label: string;
  to: string;
  renderBottomLine?: boolean;
}

export interface SidebarMenuItemWithDropdown {
  variant: "with-dropdown";
  icon: React.ReactNode;
  label: string;
  renderBottomLine?: boolean;
  dropdownItems: {
    to: string;
    label: string;
  }[];
}

export type SidebarMenuItemProps =
  | SidebarMenuItemWithoutDropdown
  | SidebarMenuItemWithDropdown;

const SidebarMenuItem = (props: SidebarMenuItemProps) => {
  const location = useLocation();

  const isSubmenuExpandedByDefault =
    props.variant === "with-dropdown"
      ? props.dropdownItems.some((item) => location.pathname.includes(item.to))
      : false;

  const [isSubmenuOpen, setIsSubmenuOpen] = useState(
    isSubmenuExpandedByDefault
  );

  const handleClick = () => setIsSubmenuOpen((prev) => !prev);

  return props.variant === "with-dropdown" ? (
    <>
      <StyledListItemButton onClick={handleClick}>
        <StyledListItemIcon>{props.icon}</StyledListItemIcon>
        <StyledListItemText primary={props.label} color="#ffffff" />
        {isSubmenuOpen ? (
          <ColoredIconWrapper color="grey">
            <ExpandLess data-testid={expandLessIconTestId} />
          </ColoredIconWrapper>
        ) : (
          <ColoredIconWrapper color="grey">
            <ExpandMore data-testid={expandMoreIconTestId} />
          </ColoredIconWrapper>
        )}
      </StyledListItemButton>
      <Collapse
        // `navigator.userAgent === "ReactSnap"` checks if react-snap is running. If so, then open tabs so they are mounted into DOM so react-snap can detect its children (Links) and create links based on that
        in={navigator.userAgent === "ReactSnap" ? true : isSubmenuOpen}
        timeout="auto"
        unmountOnExit
        data-testid={dropdownCollapsableWrapper}
      >
        <List component="div" disablePadding>
          {props.dropdownItems.map((item, index) => (
            <StyledLink key={index} to={item.to} $asSubmenuItem>
              {item.label}
            </StyledLink>
          ))}
        </List>
      </Collapse>
      {props.renderBottomLine && (
        <StyledDividerWrapper data-testid={bottomLineWrapperTestId}>
          <StyledDivider />
        </StyledDividerWrapper>
      )}
    </>
  ) : (
    <>
      <StyledLink to={props.to}>
        <StyledListItemIcon>{props.icon}</StyledListItemIcon>
        <StyledListItemText primary={props.label} />
      </StyledLink>
      {props.renderBottomLine && (
        <StyledDividerWrapper data-testid={bottomLineWrapperTestId}>
          <StyledDivider />
        </StyledDividerWrapper>
      )}
    </>
  );
};

export default SidebarMenuItem;
