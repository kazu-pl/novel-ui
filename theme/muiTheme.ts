import { createTheme } from "@mui/material/styles";

const theme = createTheme({
  palette: {
    primary: {
      light: "rgb(120, 178, 245)",
      main: "rgb(41, 114, 196)",
      dark: "rgb(35, 68, 140)",
      contrastText: "rgb(230, 238, 247)",
    },
    secondary: {
      light: "rgb(228, 152, 235)",
      main: "rgb(157, 55, 166)",
      dark: "rgb92, 35, 97)",
      contrastText: "rgb(243, 223, 245)",
    },
    // success: {},
    // warning: {},
    // error: {},
  },
});

export default theme;
