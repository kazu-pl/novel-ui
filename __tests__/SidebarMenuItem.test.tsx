import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import theme from "../theme/muiTheme";

import renderer from "react-test-renderer";

import { BrowserRouter, Link } from "react-router-dom";
import NotificationsIcon from "@mui/icons-material/Notifications";
import SidebarMenuItem, {
  SidebarMenuItemWithDropdown,
  bottomLineWrapperTestId,
  dropdownCollapsableWrapper,
} from "../src/Sidebar/SidebarMenuItem";

const Wrapper = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider theme={theme}>
      <BrowserRouter>{children}</BrowserRouter>
    </ThemeProvider>
  );
};

const dropdownItems: SidebarMenuItemWithDropdown["dropdownItems"] = [
  {
    to: "/account/avatar",
    label: "Avatar",
  },
  {
    to: "/account/settings",
    label: "Settings",
  },
  {
    to: "/account/friends",
    label: "Friends",
  },
];

describe("<SidebarMenuItem />", () => {
  it("should render correctly and match the snapshot", () => {
    const tree = renderer
      .create(
        <Wrapper>
          <SidebarMenuItem
            variant="no-dropdown"
            icon={<NotificationsIcon />}
            label={"Notifications"}
            to={"/notifications"}
          />
        </Wrapper>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render bottom line when no dropdown", () => {
    render(
      <SidebarMenuItem
        variant="no-dropdown"
        icon={<NotificationsIcon />}
        label={"Notifications"}
        to={"/notifications"}
        renderBottomLine
      />,
      {
        wrapper: Wrapper,
      }
    );

    const bottomLineWrapper = screen.getByTestId(bottomLineWrapperTestId);

    expect(bottomLineWrapper).toBeVisible();
  });

  it("should render bottom line when dropdown is used", () => {
    render(
      <SidebarMenuItem
        variant="with-dropdown"
        icon={<NotificationsIcon />}
        label={"Notifications"}
        dropdownItems={[]}
        renderBottomLine
      />,
      {
        wrapper: Wrapper,
      }
    );

    const bottomLineWrapper = screen.getByTestId(bottomLineWrapperTestId);

    expect(bottomLineWrapper).toBeVisible();
  });

  it("Dropdown should not be opened initially", () => {
    render(
      <SidebarMenuItem
        variant="with-dropdown"
        icon={<NotificationsIcon />}
        label={"Account"}
        dropdownItems={dropdownItems}
      />,
      {
        wrapper: Wrapper,
      }
    );

    const collapsableWrapper = screen.queryByTestId(dropdownCollapsableWrapper);

    expect(collapsableWrapper).toBe(null); // it should not exist
  });

  it("Dropdown should be opened when user clicks button that opens dropdown", () => {
    const label = "Account";

    render(
      <SidebarMenuItem
        variant="with-dropdown"
        icon={<NotificationsIcon />}
        label={label}
        dropdownItems={dropdownItems}
      />,
      {
        wrapper: Wrapper,
      }
    );

    const btnOpeningDropdown = screen.getByText(label);

    fireEvent.click(btnOpeningDropdown, {});

    const collapsableWrapper = screen.queryByTestId(dropdownCollapsableWrapper);

    expect(collapsableWrapper).not.toBe(null);
    expect(collapsableWrapper).toBeInTheDocument();
    expect(collapsableWrapper).toBeVisible();
  });

  it("Dropdown should NOT be opened initially when url matches only first part of url", () => {
    const label = "Account";

    render(
      <Wrapper>
        <SidebarMenuItem
          variant="with-dropdown"
          icon={<NotificationsIcon />}
          label={label}
          dropdownItems={dropdownItems}
        />
        <Link to={"/account"} data-testid="set-url-for-test" />
        <Link to="/" data-testid="reset-url" />
      </Wrapper>
    );

    const link = screen.getByTestId("set-url-for-test");
    fireEvent.click(link, {});

    const collapsableWrapper = screen.queryByTestId(dropdownCollapsableWrapper);

    expect(collapsableWrapper).toBe(null);
    expect(collapsableWrapper).not.toBeInTheDocument();
    // expect(collapsableWrapper).not.toBeVisible(); // you can't use it because collapsableWrapper is null and toBeInTheDocument() expects you pass HTML or SVG element so it can check if it's in the document or not

    // - - - - - below - is - reset - of - url - - - - -
    const resetLink = screen.getByTestId("reset-url");
    fireEvent.click(resetLink, {});
  });

  it("Dropdown should be opened INITIALLY (on page refresh) when url matches url from single dropdown item", () => {
    // First I have to render link which will set the url and after that I can render SidebarMenuItem so when it renders it already has url set
    const { rerender } = render(
      <Wrapper>
        <Link to={dropdownItems[0].to} data-testid="set-avatar-url" />
      </Wrapper>
    );
    const link = screen.getByTestId("set-avatar-url");

    fireEvent.click(link, {});

    // Now when url is already set I can render SidebarMenuItem which will be opened by default because when its internal useState will fire it wil lcheck that some of its dropdown items indeed includes location.pathname so the dropdown will be opened
    rerender(
      <Wrapper>
        <SidebarMenuItem
          variant="with-dropdown"
          icon={<NotificationsIcon />}
          label={"Avatar"}
          dropdownItems={dropdownItems}
        />
        <Link to="/" data-testid="reset-url" />
      </Wrapper>
    );

    const collapsableWrapper = screen.queryByTestId(dropdownCollapsableWrapper);

    expect(collapsableWrapper).not.toBe(null);

    expect(collapsableWrapper).toBeInTheDocument();
    expect(collapsableWrapper).toBeVisible();

    // // - - - - - below - is - reset - of - url - - - - -
    const resetLink = screen.getByTestId("reset-url");
    fireEvent.click(resetLink, {});
  });

  it("When dropdown item is clicked it should change url", () => {
    const dropdownLabel = "Account";

    render(
      <Wrapper>
        <SidebarMenuItem
          variant="with-dropdown"
          icon={<NotificationsIcon />}
          label={dropdownLabel}
          dropdownItems={dropdownItems}
        />
        <Link to="/" data-testid="reset-url" />
      </Wrapper>
    );

    const btnOpeningDropdown = screen.getByText(dropdownLabel);
    fireEvent.click(btnOpeningDropdown, {});

    const { pathname } = window.location;
    expect(pathname).toBe("/");

    const firstItemFromDropdown = screen.getByText(dropdownItems[0].label);
    fireEvent.click(firstItemFromDropdown, {});

    const { pathname: pathnameAfterClick } = window.location;
    expect(pathnameAfterClick).toBe(dropdownItems[0].to);

    // // - - - - - below - is - reset - of - url - - - - -
    const resetLink = screen.getByTestId("reset-url");
    fireEvent.click(resetLink, {});
  });
});
