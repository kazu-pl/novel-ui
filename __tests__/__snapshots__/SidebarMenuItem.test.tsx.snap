// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarMenuItem /> should render correctly and match the snapshot 1`] = `
<a
  className="sc-eCstlR dBUzxv"
  href="/notifications"
  onClick={[Function]}
>
  <div
    className="MuiListItemIcon-root-cghXcT cOXiUN MuiListItemIcon-root sc-hKgJUU juwqGD"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root-ffGNMi ufROg MuiSvgIcon-root MuiSvgIcon-fontSizeMedium"
      data-testid="NotificationsIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
      />
    </svg>
  </div>
  <div
    className="MuiListItemText-root-eslPth awSxi MuiListItemText-root sc-dlfnuX iOaXd"
  >
    <span
      className="MuiTypography-root-bWuLxc hGTtkx MuiTypography-root MuiTypography-body1 MuiListItemText-primary"
    >
      Notifications
    </span>
  </div>
</a>
`;
